airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java

```diff
@@ -77,8 +79,23 @@ public static void copyIntoTableFromStage(final JdbcDatabase database,
                                             final TyperDeduper typerDeduper)
       throws Exception {
     try {
-      stagingOperations.copyIntoTableFromStage(database, stageName, stagingPath, stagedFiles,
-          tableName, schemaName);
+      final Lock rawTableInsertLock = typerDeduper.getRawTableInsertLock(streamNamespace, streamName);
+      rawTableInsertLock.lock();
+      try {
+        stagingOperations.copyIntoTableFromStage(database, stageName, stagingPath, stagedFiles,
+            tableName, schemaName);
+      } finally {
+        rawTableInsertLock.unlock();
+      }
+
+      final AirbyteStreamNameNamespacePair streamId = new AirbyteStreamNameNamespacePair(streamName, streamNamespace);
+      if (!typerDeduperValve.containsKey(streamId)) {
+        typerDeduperValve.addStream(streamId);
+      }
+      if (typerDeduperValve.readyToTypeAndDedupe(streamId)) {
+        typerDeduper.typeAndDedupe(streamId.getNamespace(), streamId.getName(), false);
```

Comment:
Is there any issue if it is run concurrently for the same stream from multiple threads?

Reply:
typeAndDedupe handles that internally, as long as callers use `getRawTableInsertLock().lock()` correctly (https://github.com/airbytehq/airbyte/pull/29878/files#diff-7e472d19a2456a0de3d5794f56cfe9c53ed7ba1af8001bb8cacebdde488e8063R147)